File: test_validation.js
import ObscureXAgent from './agent.js';
import { writeFileSync, existsSync, unlinkSync } from 'fs';

/**
 * Validation tests for the ObscureX AI Agent
 */

async function runTests() {
  console.log('=== ObscureX Agent Validation Tests ===\n');
  
  const agent = new ObscureXAgent();
  let passed = 0;
  let failed = 0;

  // Test 1: Tool count
  console.log('Test 1: Verify all 13 tools are available');
  const tools = agent.listTools();
  if (tools.length === 13) {
    console.log('✓ PASS: All 13 tools found\n');
    passed++;
  } else {
    console.log(`✗ FAIL: Expected 13 tools, found ${tools.length}\n`);
    failed++;
  }

  // Test 2: CSV operations
  console.log('Test 2: CSV operations (add, list, remove indicators)');
  const testCSV = 'test_data.csv';
  writeFileSync(testCSV, 'date,open,high,low,close,volume\n2024-01-01,100,105,99,103,1000\n');
  
  try {
    // Add indicator
    await agent.executeTool('addTechnicalIndicator', {
      filename: testCSV,
      indicatorName: 'TEST_SMA'
    });
    
    // List indicators
    const listResult = await agent.executeTool('listTechnicalIndicators', {
      filename: testCSV
    });
    
    if (listResult.result.includes('TEST_SMA')) {
      console.log('✓ PASS: Indicator added and listed correctly');
    } else {
      console.log('✗ FAIL: Indicator not found in list');
      failed++;
    }
    
    // Remove indicator
    await agent.executeTool('removeTechnicalIndicator', {
      filename: testCSV,
      indicatorName: 'TEST_SMA'
    });
    
    const listResult2 = await agent.executeTool('listTechnicalIndicators', {
      filename: testCSV
    });
    
    if (!listResult2.result.includes('TEST_SMA')) {
      console.log('✓ PASS: Indicator removed correctly\n');
      passed++;
    } else {
      console.log('✗ FAIL: Indicator still present after removal\n');
      failed++;
    }
    
    // Cleanup
    if (existsSync(testCSV)) unlinkSync(testCSV);
  } catch (error) {
    console.log(`✗ FAIL: CSV operations error: ${error.message}\n`);
    failed++;
  }

  // Test 3: Memory operations
  console.log('Test 3: Memory operations (store, retrieve, search)');
  try {
    // Store in short-term memory
    await agent.executeTool('storeMemory', {
      key: 'test_key',
      value: { data: 'test_value' },
      metadata: { test: true }
    });
    
    // Retrieve
    const retrieved = await agent.executeTool('retrieveMemory', {
      key: 'test_key'
    });
    
    if (retrieved.success && retrieved.result.value.data === 'test_value') {
      console.log('✓ PASS: Short-term memory store/retrieve works');
    } else {
      console.log('✗ FAIL: Short-term memory retrieve failed');
      failed++;
    }
    
    // Search
    const searchResult = await agent.executeTool('searchMemory', {
      query: 'test_key'
    });
    
    if (searchResult.result.count > 0) {
      console.log('✓ PASS: Short-term memory search works');
      passed++;
    } else {
      console.log('✗ FAIL: Short-term memory search failed');
      failed++;
    }
    
    // Long-term memory
    await agent.executeTool('storeLongTermMemory', {
      key: 'lt_test',
      value: { persistent: true }
    });
    
    const ltRetrieved = await agent.executeTool('retrieveLongTermMemory', {
      key: 'lt_test'
    });
    
    if (ltRetrieved.success) {
      console.log('✓ PASS: Long-term memory works\n');
      passed++;
    } else {
      console.log('✗ FAIL: Long-term memory failed\n');
      failed++;
    }
  } catch (error) {
    console.log(`✗ FAIL: Memory operations error: ${error.message}\n`);
    failed++;
  }

  // Test 4: Python module listing
  console.log('Test 4: Python module operations');
  try {
    const modules = await agent.executeTool('listPythonModules', {});
    
    if (modules.success && modules.result.length > 0) {
      console.log(`✓ PASS: Listed ${modules.result.length} Python modules\n`);
      passed++;
    } else {
      console.log('✗ FAIL: No Python modules found\n');
      failed++;
    }
  } catch (error) {
    console.log(`✗ FAIL: Python module listing error: ${error.message}\n`);
    failed++;
  }

  // Test 5: Pipeline generation (template-based, no API key needed)
  console.log('Test 5: ML Pipeline generation (template-based)');
  try {
    const generated = await agent.executeTool('generateMLPipeline', {
      existingCode: null,
      prompt: 'Create a test pipeline using CatBoost'
    });
    
    if (generated.success && generated.result.code.includes('CatBoost')) {
      console.log('✓ PASS: Pipeline generation works');
      console.log(`  Generated file: ${generated.result.filename}\n`);
      passed++;
      
      // Cleanup generated file
      if (existsSync(generated.result.filename)) {
        unlinkSync(generated.result.filename);
      }
    } else {
      console.log('✗ FAIL: Pipeline generation failed\n');
      failed++;
    }
  } catch (error) {
    console.log(`✗ FAIL: Pipeline generation error: ${error.message}\n`);
    failed++;
  }

  // Test 6: ML Pipeline testing
  console.log('Test 6: ML Pipeline testing');
  const simplePipeline = `
import numpy as np
from sklearn.metrics import mean_squared_error

y_true = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
y_pred = np.array([1.1, 2.1, 3.1, 4.1, 5.1])

mse = mean_squared_error(y_true, y_pred)
print(f'MSE: {mse}')
`;
  
  try {
    const testResult = await agent.executeTool('testMLPipeline', {
      pythonCode: simplePipeline
    });
    
    if (testResult.result.success && testResult.result.mse !== null) {
      console.log(`✓ PASS: Pipeline testing works (MSE: ${testResult.result.mse})\n`);
      passed++;
    } else {
      console.log('✗ FAIL: Pipeline testing failed or MSE not extracted\n');
      failed++;
    }
  } catch (error) {
    console.log(`✗ FAIL: Pipeline testing error: ${error.message}\n`);
    failed++;
  }

  // Summary
  console.log('=== Test Summary ===');
  console.log(`Passed: ${passed}`);
  console.log(`Failed: ${failed}`);
  console.log(`Total: ${passed + failed}`);
  
  if (failed === 0) {
    console.log('\n✓✓✓ All tests passed! ✓✓✓');
  } else {
    console.log(`\n⚠ ${failed} test(s) failed`);
  }
  
  return failed === 0;
}

// Run tests
runTests()
  .then(success => {
    process.exit(success ? 0 : 1);
  })
  .catch(error => {
    console.error('Test suite failed:', error);
    process.exit(1);
  });

File: test_validation.py
#!/usr/bin/env python3
"""
Test validation for ObscureX Agent
"""

import asyncio
import os
import sys
from obscurex import ObscureXAgent


async def run_tests():
    """Run validation tests for the agent."""
    print('=== ObscureX Agent Validation Tests (Python) ===\n')
    
    agent = ObscureXAgent()
    passed = 0
    failed = 0
    total = 0
    
    # Test 1: Verify all tools are available
    print('Test 1: Verify all tools are available')
    total += 1
    tools = await agent.list_tools()
    if len(tools) >= 13:  # Should have at least 13 tools
        print(f'✓ PASS: Found {len(tools)} tools (expected at least 13)\n')
        passed += 1
    else:
        print(f'✗ FAIL: Only found {len(tools)} tools\n')
        failed += 1
    
    # Test 2: CSV operations
    print('Test 2: CSV operations (add, list, remove indicators)')
    total += 1
    try:
        # Create test file
        with open('test_data.csv', 'w') as f:
            f.write('date,open,high,low,close,volume\n')
            f.write('2024-01-01,100,105,99,103,1000\n')
        
        # Add indicator
        await agent.execute_tool('addTechnicalIndicator', {
            'filename': 'test_data.csv',
            'indicatorName': 'TEST_IND'
        })
        
        # List indicators
        result = await agent.execute_tool('listTechnicalIndicators', {
            'filename': 'test_data.csv'
        })
        
        if 'TEST_IND' in result['result']:
            print('✓ PASS: Indicator added and listed correctly')
            
            # Remove indicator
            await agent.execute_tool('removeTechnicalIndicator', {
                'filename': 'test_data.csv',
                'indicatorName': 'TEST_IND'
            })
            
            result = await agent.execute_tool('listTechnicalIndicators', {
                'filename': 'test_data.csv'
            })
            
            if 'TEST_IND' not in result['result']:
                print('✓ PASS: Indicator removed correctly\n')
                passed += 1
            else:
                print('✗ FAIL: Indicator not removed\n')
                failed += 1
        else:
            print('✗ FAIL: Indicator not added\n')
            failed += 1
        
        # Cleanup
        if os.path.exists('test_data.csv'):
            os.remove('test_data.csv')
    except Exception as e:
        print(f'✗ FAIL: {e}\n')
        failed += 1
    
    # Test 3: Memory operations
    print('Test 3: Memory operations (store, retrieve, search)')
    total += 1
    try:
        # Store
        await agent.execute_tool('storeMemory', {
            'key': 'test_mem',
            'value': 'test_value',
            'metadata': {}
        })
        
        # Retrieve
        result = await agent.execute_tool('retrieveMemory', {
            'key': 'test_mem'
        })
        
        if result['result']['value'] == 'test_value':
            print('✓ PASS: Short-term memory store/retrieve works')
            
            # Search
            result = await agent.execute_tool('searchMemory', {
                'query': 'test_mem'
            })
            
            if result['result']['count'] > 0:
                print('✓ PASS: Short-term memory search works')
                
                # Long-term memory
                await agent.execute_tool('storeLongTermMemory', {
                    'key': 'test_lt',
                    'value': 'long_term_value',
                    'metadata': {}
                })
                
                result = await agent.execute_tool('retrieveLongTermMemory', {
                    'key': 'test_lt'
                })
                
                if result['result']['value'] == 'long_term_value':
                    print('✓ PASS: Long-term memory works\n')
                    passed += 1
                else:
                    print('✗ FAIL: Long-term memory retrieve failed\n')
                    failed += 1
            else:
                print('✗ FAIL: Memory search failed\n')
                failed += 1
        else:
            print('✗ FAIL: Memory retrieve failed\n')
            failed += 1
    except Exception as e:
        print(f'✗ FAIL: {e}\n')
        failed += 1
    
    # Test 4: Python module operations
    print('Test 4: Python module operations')
    total += 1
    try:
        result = await agent.execute_tool('listPythonModules', {})
        if result['success'] and len(result['result']) > 0:
            print(f'✓ PASS: Listed {len(result["result"])} Python modules\n')
            passed += 1
        else:
            print('✗ FAIL: Failed to list modules\n')
            failed += 1
    except Exception as e:
        print(f'✗ FAIL: {e}\n')
        failed += 1
    
    # Test 5: AI features (skip if no API key)
    print('Test 5: ML Pipeline generation (AI-powered)')
    total += 1
    if not agent.ai_service.is_available():
        print('⚠ SKIP: AI API key not available (test requires Anthropic Claude API)\n')
    else:
        try:
            result = await agent.execute_tool('generateMLPipeline', {
                'existingCode': None,
                'prompt': 'Create a simple linear regression pipeline'
            })
            if result['success']:
                print(f'✓ PASS: Pipeline generated: {result["result"]["filename"]}\n')
                passed += 1
            else:
                print(f'✗ FAIL: {result.get("error")}\n')
                failed += 1
        except Exception as e:
            print(f'✗ FAIL: {e}\n')
            failed += 1
    
    # Summary
    print('=== Test Summary ===')
    print(f'Passed: {passed}')
    print(f'Failed: {failed}')
    print(f'Total: {total}')
    
    if failed > 0:
        print(f'\n⚠ {failed} test(s) failed')
        return 1
    else:
        print('\n✓ All tests passed!')
        return 0


if __name__ == '__main__':
    try:
        exit_code = asyncio.run(run_tests())
        sys.exit(exit_code)
    except KeyboardInterrupt:
        print('\nInterrupted by user')
        sys.exit(0)
    except Exception as e:
        print(f'\nError: {e}')
        import traceback
        traceback.print_exc()
        sys.exit(1)
